azure/ConsiderationsForServiceDesign.md

```diff
@@ -229,6 +229,33 @@ An important distinction between RELO and status monitor LROs is that there is a
+
+There are two types of errors returned from your service and customers handle them differently.
+- Usage errors where the customer is calling your API incorrectly.  The customer can easily make these errors go away by fixing their code.  We expect most usage errors to be found during testing.
+- Runtime errors that can't be prevented by the customer and need to be recovered from.  Some runtime errors like `429` throttling will be handled automatically by client libraries, but most will be situations like a `409` conflict requiring knowledge about the customer's application to remedy.
+
+We provide appropriate [HTTP status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) for customers to handle errors generically and error code strings in our common error schema and the `x-ms-error-code` header for customers to handle errors specifically.  As an example, consider what a customer would do when trying to get the properties of a Storage blob:
+- A `404` status code tells them the blob doesn't exist and the customer can report the error to their users
+- A `BlobNotFound` or `ContainerNotFound` error code will tell them why the blob doesn't exist so they can take steps to recreate it
+
+The common error schema in the Guidelines allows nested details and inner errors that have their own error codes, but the top-level error code is the most important.  The HTTP status code and the top-level error code are the only part of your error that we consider part of your API contract that follows the same compatibility requirements as the rest of your API.  Importantly, this means you **can not change the HTTP status code or top-level error code for an API in GA'ed service version**.  You can only return new status codes and error codes in future API versions if customers make use of new features that trigger new classes of errors.  Battle tested error handling is some of the hardest code to get right and we can't break that for customers when they upgrade to the latest version.  The rest of the properties in your error like `message`, `details`, etc., are not considered part of your API contract and can change to improve the diagnosability of your service.
+
+We also return the top-level error code as the `x-ms-error-code` header so client libraries have the ability to automatically retry requests when possible without having to parse a JSON payload.  We recommend unique error codes like `ContainerBeingDeleted` for every distinct recoverable error that can occur, but suggest reusing common error codes like `InvalidHeaderValue` for usage errors where a descriptive error message is more important for resolving the problem.  The Storage [Common](https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes) and [Blob](https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-error-codes) error codes are a good starting point if you're looking for examples.
```

Comment:
Would it be unreasonable to simply use the status code name as the code for usage type errors, e.g. "BadRequest"?  Said differently, if an error is not expected to be handled programmatically, then the developer will have the benefit of the error message -- which should be as specific as possible -- to determine the cause of the error.  Is "InvalidHeader" any better that "BadRequest" given that this situation will not be resolved programmatically? Is it even desirable to have more than one error code for "BadRequest" if none can be handled programmatically?

This is the part that still confuses me about the two error types.  If both types get named error codes (other than the status code name), then how can I tell which are expected to be "recoverable"?

Reply:
Mike: I don't think that returning BadRequest for 400 is useful.  The [HTTP spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html) mandates that all HTTP responses have a status code and a "reason phrase". For 400, The reason phrase will already have "Bad Request" in it. Adding it again in the error message isn't useful and bloats what's sent over the wire. However, having the reason for the bad request is useful in the error code/message.

Reply:
I'm going to punt this to https://github.com/microsoft/api-guidelines/issues/312 where we can discuss whether we only want `400`/`BadRequest` or to go further with `400`/`InvalidValue`, `400`/`MissingRequiredValue`, etc.  I personally like the "more specific generic error codes" and see it as a happy medium between the too generic `BadRequest` and too specific `ContainerNameMustHaveBetween3And32Characters`.  I don't think we can tell people to Goldilocks their APIs until they're just right though, so https://github.com/microsoft/api-guidelines/issues/312 as an example is probably our best bet.

As for documenting recoverable errors, I'm going to punt that to [my side project](https://microsoft-my.sharepoint.com/:p:/p/teglaza/EeTYvK4JCJNIjtA1BLufg34B4MROlR-kaUACj3HqXidpNQ?e=Resebh) (see slide 34 in particular).

Reply:
Agree. Let's provide this updated guidance and move this to #312 .